### PR TITLE
Refactor Active Record adapters to have a similar internal interface

### PR DESCRIPTION
**Pull request by casperisfine**
Adapters have very inconsistent internal APIs to perform queries.

This refactoring tries to improve consistency with a common provite API for all of them.

Abstract methods:

- `perform_query`: the only method where an adapter should perform a query. It returns a native, adapter specific result object. Does not apply query transformations. Does not check for writes. Receive the raw connection object to use as argument.

- `cast_result`: receives the native result object and returns a generic `ActiveRecord::Result`.

- `affected_rows`: receives the native result object and returns the number of affected rows.

By just implementing these 3 methods all adapters automatically get:

- `raw_execute`: Perform the query and returns a native result object. Also emits a `sql.active_record` event.

- `raw_exec_query`: same as `raw_execute` but returns an `ActiveRecord::Result`.

- `internal_exec_query`: same as `raw_exec_query` but check for writes and apply query transformations.

- `internal_execute`: same as `internal_exec_query` but returns the native, adapter specific, result object.

With this increased consistency, we can now reduce the amount of duplicated code in every adapter. There's some room for further improvements but I tried to not go too far all at once.

Also previously some adapters had a block based query interface that allowed to eagerly clear the native result object.

It may make sense to bring that capability back in a consistent way, but short term I opted for consistency.

FYI: @matthewd @adrianna-chang-shopify 

Also FYI: @yahonda @aidanharan (I don't expect breakage to either of your adapters but I haven't tested them to be honest)

### CODE REVIEW

**Comment by casperisfine** [comment]
Discussed this with @matthewd, he had 3 concerns:

  - `PostgresAdapter` having an inverted `with_raw_connection / log` sequence is weirdPreviously it would only do so if prepared statements were enabled, now always.
  - The above issue prevent removing `with_raw_connection / log` from `raw_execute`.
  -  Bit concerned about losing the block form that allow eagerly freeing the result object.

I just pushed a commit that fixes the first concern, which open the door to fix the second one.

As for the third concern, I think it's possible to bring that capability back, I just need to have a look.

**Comment by casperisfine** [comment]
Alright, I think this is good to go now. If there are some extra comments, we can iterate on it.

**Comment by matthewd** [comment]
> Alright, I think this is good to go now. If there are some extra comments, we can iterate on it.

:+1:

> the only adapter that was benefiting from using a block was SQlite3

Ah, I see you did keep the pg `clear` and just moved it to `cast_result`. I don't _love_ the fact that downgrades it from an `ensure` to something the caller is responsible for (and technically missing from `affected_rows`, though I'd be quite surprised if that ever saw a result large enough to care about), but overall probably fine.

On a similar note, I think it makes sense that `cast_result` destructively consumes the result it's given, but we should probably remember to note that explicitly if we ever formalize this API definition.

---

I _do_ love the `raw_execute` / `perform_query` simplification 😍

**Comment by byroot** [comment]
Right. I think it really wouldn't be hard for the `*execute` methods to optionally take a block, and when they have one, yield the result and then clear it.

**Comment by byroot** [comment]
>  though I'd be quite surprised if that ever saw a result large enough to care about), but overall probably fine.

That could be the difference between triggering the GC more or less often with doing a large `find_each`. Especially since the default `RUBY_GC_MALLOC_LIMIT_MAX` is only `32MB`, so eagerly freeing these stuff might make a big difference in some cases.

File: activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb [hunk 1]
```diff
@@ -170,8 +132,61 @@ def cancel_any_running_query
...
+                      @statements.delete sql_key(sql)
+                    end
+                    retry
+                  end
+                end
+
+                raise
+              end
+            elsif without_prepared_statement?(binds)
+              raw_connection.async_exec(sql)
```
Comment (robinator) [inline activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:160]: This change is discarding the `binds` for queries where `prepare` is false, which is currently breaking `good_job` in many places.  Is this change intentional or a regression?  I believe binds should be included regardless of whether prepare is true but I am no expert!

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:160]: Also this branch isn't entrered unless binds are empty / nil, so I don't see how it could break anything.

Comment (robinator) [reply activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:160]: The error appeared [here](https://github.com/bensheldon/good_job/blob/a61fc09cdc06ba89a9615da90778459f13a837a8/app/charts/good_job/scheduled_by_queue_chart.rb#L33) for us.  Currently, this branch is entered as long as `!prepared_statements` returns true (regardless of how binds is set)

```
def without_prepared_statement?(binds)
  !prepared_statements || binds.nil? || binds.empty?
end
```

File: activerecord/test/cases/adapters/postgresql/connection_test.rb [hunk 2]
```diff
@@ -130,8 +130,11 @@ def test_schema_names_logs_name
       def test_statement_key_is_logged
         bind = Relation::QueryAttribute.new(nil, 1, Type::Value.new)
         @connection.exec_query("SELECT $1::integer", "SQL", [bind], prepare: true)
```
Comment (robinator) [inline activerecord/test/cases/adapters/postgresql/connection_test.rb:132]: @byroot You could duplicate this test and set `prepare: false` (or leave out of args) to replicate the issue.

Comment (byroot) [reply activerecord/test/cases/adapters/postgresql/connection_test.rb:132]: No, it's not as simple as that unfortunately.

File: activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb (hunk 1 above)
Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/postgresql/database_statements.rb:160]: https://github.com/rails/rails/pull/52580
